### PR TITLE
typing interface

### PR DESCRIPTION
**Pull request by Tishka17**
#25

### CODE REVIEW

**Comment by Tishka17** [comment]
@sobolevn review pls

File: dataclass_factory/factory.py [hunk 1]
```diff
@@ -38,49 +39,56 @@ def serializer(self, class_: Type):
...
-        self.schemas: Dict[Type, Schema] = defaultdict(lambda: copy(self.default_schema))
+        self.schemas: Dict[Type, Schema] = {}
         if schemas:
             self.schemas.update({
                 type_: merge_schema(schema, self.default_schema)
                 for type_, schema in schemas.items()
             })
 
-    def schema(self, class_: Type) -> Schema:
+    def schema(self, class_: Type[T]) -> Schema:
```
Comment (sobolevn) [inline dataclass_factory/factory.py]: It returns `Schema[Any]`, should it work this way?

File: dataclass_factory/factory.py [hunk 2]
```diff
@@ -38,49 +39,56 @@ def serializer(self, class_: Type):
...
+            )
             self.schemas[class_] = schema
         return schema
 
-    def parser(self, class_: Type) -> Parser:
+    def parser(self, class_: Type[T]) -> Parser[T]:
         return self._parser_with_stack(class_, StackedFactory(self))
 
-    def _parser_with_stack(self, class_: Type, stacked_factory: StackedFactory) -> Parser:
+    def _parser_with_stack(self, class_: Type[T], stacked_factory: StackedFactory) -> Parser:
```
Comment (sobolevn) [inline dataclass_factory/factory.py]: This line is inconsistent with the previous one: `def parser(self, class_: Type[T]) -> Parser[T]:`

Comment (sobolevn) [reply dataclass_factory/factory.py]: But it looks like totally the same.

File: dataclass_factory/factory.py [hunk 3]
```diff
@@ -38,49 +39,56 @@ def serializer(self, class_: Type):
...
         if not schema.parser:
             schema.parser = create_parser(stacked_factory, schema, self.debug_path, class_)
         return schema.parser
 
-    def serializer(self, class_: Type) -> Serializer:
+    def serializer(self, class_: Type[T]) -> Serializer[T]:
         return self._serializer_with_stack(class_, StackedFactory(self))
 
-    def _serializer_with_stack(self, class_: Type, stacked_factory: StackedFactory) -> Serializer:
+    def _serializer_with_stack(self, class_: Type[T], stacked_factory: StackedFactory) -> Serializer:
```
Comment (sobolevn) [inline dataclass_factory/factory.py]: The same here: `def serializer(self, class_: Type[T]) -> Serializer[T]:` is not in sync with `-> Serializer:`

File: dataclass_factory/factory.py [hunk 4]
```diff
@@ -38,49 +39,56 @@ def serializer(self, class_: Type):
...
         if not schema.serializer:
             schema.serializer = create_serializer(stacked_factory, schema, self.debug_path, class_)
         return schema.serializer
 
-    def load(self, data: Any, class_: Type):
+    def load(self, data: Any, class_: Type[T]) -> T:
         return self.parser(class_)(data)
 
-    def dump(self, data: Any, class_: Type = None):
+    def dump(self, data: T, class_: Type[T] = None) -> Any:
```
Comment (sobolevn) [inline dataclass_factory/factory.py]: It returns `Union[List[Any], Dict[Any]]` I guess (not sure). We can narrow the type.

File: dataclass_factory/serializers.py [hunk 5]
```diff
@@ -26,37 +25,37 @@ def serialize(data):
...
     return lambda data: [serializer(x) for x in data]
 
 
-def get_tuple_serializer(serializers) -> Serializer:
+def get_tuple_serializer(serializers) -> Serializer[List]:
     return lambda data: [serializer(x) for x, serializer in zip(data, serializers)]
 
 
-def get_collection_any_serializer() -> Serializer:
+def get_collection_any_serializer() -> Serializer[List]:
```
Comment (sobolevn) [inline dataclass_factory/serializers.py]: Is `List[Any]` ok here?

File: dataclass_factory/serializers.py [hunk 6]
```diff
@@ -26,37 +25,37 @@ def serialize(data):
...
 
-def get_dict_serializer(serializer):
+def get_dict_serializer(serializer: Any) -> Serializer[Dict]:
     return lambda data: {
         k: serializer(v) for k, v in data.items()
     }
 
 
-def lazy_serializer(factory):
+def lazy_serializer(factory) -> Serializer:
```
Comment (sobolevn) [inline dataclass_factory/serializers.py]: Is it possible to type `Serializer`?

File: dataclass_factory/factory.py (hunk 4 above)
Comment (Tishka17) [reply dataclass_factory/factory.py]: No. Return type can be different for different type. E.g. `dump(1)==1`

File: dataclass_factory/serializers.py (hunk 5 above)
Comment (Tishka17) [reply dataclass_factory/serializers.py]: Yes, it is.  i'll specify explicitly

File: dataclass_factory/serializers.py (hunk 6 above)
Comment (Tishka17) [reply dataclass_factory/serializers.py]: No. In this case created serializer can serialize any data, beacuse it find concrete serializer in every call

File: dataclass_factory/factory.py (hunk 4 above)
Comment (sobolevn) [reply dataclass_factory/factory.py]: 👍

**Pull request merged**